nn/nnclient.py

```diff
@@ -0,0 +1,305 @@
+        a parse tree of Icelandic text """
+
+    port = Settings.NN_PARSING_PORT
+    host = Settings.NN_PARSING_HOST
+    verb = "parse"
+
+    @classmethod
+    def _processResponse(cls, instance, sent):
+        """ Process the response from a single sentence """
+        try:
+            instance["scores"] = max([float(score) for score in instance["scores"]])
+        except TypeError:
+            # Score is not iterable
+            pass
+        return instance
+
+    @classmethod
+    def _instances_to_ptrees(cls, insts, sents):
+        """ Transforms list of result dicts of flat parse trees
+            into a list of dicts of parsed tree structures """
+        for (inst, sent) in zip(insts, sents):
+            parse_toks = inst["outputs"]
+            tree, p_result = nntree.parse_tree_with_text(parse_toks, sent)
+            inst["outputs"] = tree
+
+            if Settings.DEBUG:
+                print("Received parse tokens from nnserver:", parse_toks)
+                print("ParseResult: ", p_result)
+                print("Parsed by nntree into:")
+                tree.pprint()
+                if p_result != ParseResult.SUCCESS:
+                    print(
+                        "NnParse not successful for input: '{text}'".format(text=sent)
+                    )
+                    print("ParseResult: {result}".format(result=p_result))
+                    print("Output: {parse_toks}".format(parse_toks=parse_toks))
+
+    @classmethod
+    def request_text(cls, text, flat=False):
+        """ Request neural network output for contiguous text """
+        sents = cls._normalizeText(text)
+        results = cls._request(sents)
+        if not flat:
+            cls._instances_to_ptrees(results, sents)
+        return results
+
+    @classmethod
+    def request_sentence(cls, text, flat=False):
+        """ Request neural network output for a single sentence """
+        if "\n" in text:
+            single_sentence = text.split("\n")[0]
+        else:
+            single_sentence = text
+        toks = cls._normalizeSentence(single_sentence)
+        single_sentence = " ".join(toks)
+        sents = [single_sentence]
+        results = cls._request(sents)
+        if results is None:
+            return results
+        if not flat:
+            cls._instances_to_ptrees(results, sents)
+        return results[0]
+
+    @classmethod
+    def _normalizeText(cls, text):
+        """ Preprocess text and normalize for parsing network """
+        pgs = text.split("\n")
+        normalized_pgs = [
+            [tok.txt for tok in list(bintokenizer.tokenize(pg))
+                if BIN_Token.is_understood(tok)
+            ] for pg in pgs]
+        normalized_pgs = [
+            " ".join([tok for tok in npg if tok]) for npg in normalized_pgs
+        ]
+        return normalized_pgs
+
+    @classmethod
+    def _normalizeSentence(cls, single_sentence):
+        """ Preprocess text and normalize for parsing network """
```

Comment:
Factor this out to a helper function and call it everywhere this pattern occurs?